### PR TITLE
Adding new device interface for post comm creation op

### PR DESCRIPTION
**Pull request by yfguo**
## Pull Request Description

Some collective related init code for the comm need to be called at the end of comm creation. Adding this interface enables that without requiring refactoring the CH4 comm creation hook and separating current rankmap code. As I mentioned in the call this morning, the refactoring breaks CH3 and would require implementing the av table related stuff in CH3 to replace VCs.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
So the idea is this hook would be used to setup collective infrastructure on the new communicator? Is there any example?

**Comment by raffenet** [comment]
What is the requirement for this hook to be useful? That the communicator can be used for collective communication before returning to the user? Just trying to understand the whole picture.

**Comment by yfguo** [comment]
@raffenet One example is that the device level collective selection logic would query the newly created communication for information like subcomms, the results selection parameters need to be saved into the communication. Some of these queries can only be done after `MPID_Comm_create_hook` is called, therefore it would be nontrivial to implementing this without adding the `MPID_Comm_create_hook_post`. I think @zhenggb72 have more information about this.

**Comment by zhenggb72** [comment]
Yes, we have a use case that we need to implement some communicator specific collective selection optimization after the communicator is fully created, the device level collective algorithm selection may requires information about the communicator's subcomms to prune the search space.

**Comment by raffenet** [comment]
Could we add device hooks in `MPIR_Coll_comm_init` to the same end? E.g. `MPID_Coll_comm_init_hook`? I think it would improve understanding.

**Comment by yfguo** [comment]
@zhenggb72 @raffenet I am OK with the `MPID_Coll_comm_init_hook`. There is already a `MPIR_Coll_comm_init`. Maybe we can incorporate that as well. Let me update the PR.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch3/tcp

**Comment by yfguo** [comment]
@zhenggb72 Can you try this new PR with Intel's code to see if it is sufficient for the collective initialization?

**Comment by zhenggb72** [comment]
This PR is good. I can implement the collective initialization on top of that.

**Comment by yfguo** [comment]
rebased and kick the test again.
test:jenkins/ch4/ofi
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
This makes sense to me, but there was some discussion about #4214 possibly addressing this problem in a different way? Unless I am mistaken, we should make sure we are on the same page before merging.

**Comment by raffenet** [comment]
> This makes sense to me, but there was some discussion about #4214 possibly addressing this problem in a different way? Unless I am mistaken, we should make sure we are on the same page before merging.

After discussion, my comment is inaccurate. This is solving a different issue, so I'm going to approve and merge.

File: src/mpi/comm/commutil.c [hunk 1]
```diff
@@ -310,10 +310,6 @@ static int MPIR_Comm_commit_internal(MPIR_Comm * comm)
     mpi_errno = MPID_Comm_create_hook(comm);
     MPIR_ERR_CHECK(mpi_errno);
 
-    /* Create collectives-specific infrastructure */
```
Comment (raffenet) [inline src/mpi/comm/commutil.c]: @yfguo can we not delete this comment? When moving this call to the new location? I think it is still helpful.

**Comment by zhenggb72** [comment]
Is this PR good to go?

**Comment by raffenet** [comment]
> Is this PR good to go?

My review comment has not been addressed.

**Comment by yfguo** [comment]
rebased and ready for merge.
test:jenkins/ch4/ofi
test:jenkins/ch3/tcp

File: src/mpi/comm/commutil.c (hunk 1 above)
Comment (raffenet) [reply src/mpi/comm/commutil.c]: This comment is now out of place. Please move it to the new location of MPIR_Coll_comm_init

**Comment by yfguo** [comment]
Sorry, wasn't sure what was I thinking of doing that change.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch3/tcp

**Pull request merged**